Update GA to version 4

MarekB01 opened the pull request:
https://github.com/BroadleafCommerce/QA/issues/4848

Updated to GA4.
Created a custom thymeleaf processor for new version of GA

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

created custom processor

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

created custom processor, Refactored a code

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Refactored code

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Upgraded to GA4. Migrated from analytics.js to gtag.js which uses google tag manager

oleksiimiroshnyk commented on core/broadleaf-framework-web/src/main/java/org/broadleafcommerce/core/web/processor/GoogleAnalytics4Processor.java:
```diff
@@ -0,0 +1,332 @@
+     * in the given <b>order</b>.
+     */
+    protected String getTransactionJs(Order order, String trackerPrefix) {
+        StringBuffer sb = new StringBuffer();
```
guess StringBuilder is better ? Buffer is synch which is not needed here

oleksiimiroshnyk commented on core/broadleaf-framework-web/src/main/java/org/broadleafcommerce/core/web/processor/GoogleAnalytics4Processor.java:
```diff
@@ -0,0 +1,332 @@
+
+    @Override
+    public BroadleafTemplateModel getReplacementModel(String tagName, Map<String, String> tagAttributes, BroadleafTemplateContext context) {
+        StringBuffer sb = new StringBuffer();
```
builder instead of buffer

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Upgraded to GA4. Migrated from analytics.js to gtag.js which uses google tag manager. Refactored code

oleksiimiroshnyk commented on core/broadleaf-framework-web/src/main/java/org/broadleafcommerce/core/web/processor/GoogleUniversalAnalyticsProcessor.java:
```diff
@@ -111,10 +115,12 @@ public BroadleafTemplateModel getReplacementModel(String tagName, Map<String, St
         StringBuffer sb = new StringBuffer();
```
one more buffer

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Upgraded to GA4. Migrated from analytics.js to gtag.js which uses google tag manager. Refactored code.

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Upgraded to GA4. Migrated from analytics.js to gtag.js which uses google tag manager. Refactored code.

riteshadhikari17 commented:
@MarekB01 @oleksiimiroshnyk I don't know how to test this. Have you guys done any testing? If you guys are confident about this, you can simply merge it.

Jerry77oz commented:
@MarekB01 this looks like we're good to merge this, anything you think we could test before we do?

MarekB01 committed:
BroadleafCommerce/QA-4848_update_google_analytics

Refactored code, renamed variables and properties

Commit:
BroadleafCommerce/QA#4848
- Migrated google analytics to gtag
- added support for tag id for universal analytics -> ga4 migration:
https://support.google.com/analytics/answer/9310895#gtagjs-enable-basic&zippy=%2Cin-this-article
https://developers.google.com/analytics/devguides/migration/ecommerce/gtagjs-ua-to-ga4
https://developers.google.com/analytics/devguides/migration/ecommerce/analyticsjs-before-you-begin

oleksiimiroshnyk commented:
used docs for migration:
https://developers.google.com/analytics/devguides/migration/ecommerce/analyticsjs-before-you-begin\
https://developers.google.com/analytics/devguides/migration/ecommerce/gtagjs-ua-to-ga4
https://support.google.com/analytics/answer/9310895#gtagjs-enable-basic&zippy=%2Cin-this-article
https://support.google.com/analytics/answer/10759417
https://support.google.com/analytics/answer/9744165#zippy=%2Cin-this-article

new property for tag id: googleAnalytics4.tagIdForProperty

Commit:
BroadleafCommerce/QA#4848
formatting
